packages/host/app/components/ai-assistant/formatted-message.gts

```diff
@@ -249,25 +249,93 @@ class HtmlDidUpdate extends Modifier<HtmlDidUpdateSignature> {
 }
 
 function parseHtmlContent(htmlString: string): HtmlTagGroup[] {
-  const result: HtmlTagGroup[] = [];
-
-  // Regular expression to match <pre> tags and their content
-  const regex = /(<pre[\s\S]*?<\/pre>)|([\s\S]+?)(?=<pre|$)/g;
-
-  let match;
-  while ((match = regex.exec(htmlString)) !== null) {
-    if (match[1]) {
-      // This is a code block (pre tag)
-      result.push({
-        type: 'pre_tag',
-        content: match[1],
-      });
-    } else if (match[2] && match[2].trim() !== '') {
-      // This is non <pre> tag HTML
-      result.push({
-        type: 'non_pre_tag',
-        content: match[2],
-      });
+  let result: HtmlTagGroup[] = [];
+  let tagStack: { tag: string; startPos: number }[] = [];
+  let currentPosition = 0;
+
+  let findNextTag = (
+    pos: number,
+  ): { type: 'open' | 'close'; tag: string; position: number } | null => {
+    let currentPos = pos;
+    while (currentPos < htmlString.length) {
+      let openTag = htmlString.indexOf('<', currentPos);
+      if (openTag === -1) return null;
+
+      if (htmlString.startsWith('<<<<<<<', openTag)) {
+        let endMarker = htmlString.indexOf('>>>>>>>', openTag);
+        if (endMarker === -1) {
+          currentPos = openTag + 7; // length of '<<<<<<<'
+          continue;
+        }
+        // Skip past the entire search/replace block
+        currentPos = endMarker + 7; // length of '>>>>>>>'
```

Comment:
I am confused why you are dealing with search/replace blocks here. Parsing of the search/replace blocks is already done `extractCodeData`